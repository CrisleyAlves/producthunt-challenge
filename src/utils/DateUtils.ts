import { POST_ORDER_TYPE } from "../config/constants";

/**
 * Returns a new Date object representing the previous day of the given date.
 *
 * @param date - The original date. Can be a Date object or a value accepted by the Date constructor.
 * @returns A new Date object set to one day before the provided date.
 */
export function previousDay(date: Date | string): Date {
  const copy = new Date(date);
  copy.setDate(copy.getDate() - 1);
  return copy;
}

/**
 * Generates parameters for fetching posts based on the specified order and post date.
 *
 * @param order - The order type as a string (e.g., 'asc', 'desc').
 * @param postDate - The date string representing the reference post date.
 * @returns An object containing:
 *   - orderType: The resolved order type value from POST_ORDER_TYPE.
 *   - selectedDate: The Date object created from the provided postDate.
 *   - yesterday: The Date object representing the day before selectedDate.
 */
export function getPostsParams(order: string, postDate: string) {
  const orderType = POST_ORDER_TYPE[order.toUpperCase()].value;
  const selectedDate = new Date(postDate);
  const yesterday = previousDay(selectedDate);

  return {
    orderType,
    selectedDate,
    yesterday,
  };
}
