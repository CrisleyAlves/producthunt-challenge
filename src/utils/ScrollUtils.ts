/**
 * @description Uses `window.requestAnimationFrame` to ensure the scroll action
 * is performed in the next animation frame for better performance.
 */
export function scrollToTop() {
  window.requestAnimationFrame(() => {
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  });
}
