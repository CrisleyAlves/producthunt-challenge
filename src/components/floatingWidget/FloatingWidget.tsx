import React from "react";
import { faArrowUp } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { scrollToTop } from "../../utils/ScrollUtils";

export default function FloatingWidget() {
  return (
    <button
      className="floating-widget"
      onClick={() => scrollToTop()}
      aria-label="Scroll to top"
    >
      <FontAwesomeIcon className="icon" icon={faArrowUp} />
    </button>
  );
}
