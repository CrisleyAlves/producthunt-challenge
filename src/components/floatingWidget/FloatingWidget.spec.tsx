import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";

import * as ScrollUtils from "../../utils/ScrollUtils";
import FloatingWidget from "./FloatingWidget";

describe("FloatingWidget", () => {
  it("renders the floating widget button", () => {
    const { container } = render(<FloatingWidget />);

    expect(container).toMatchSnapshot();
  });

  it("calls scrollToTop when button is clicked", () => {
    const scrollToTopSpy = vi
      .spyOn(ScrollUtils, "scrollToTop")
      .mockImplementation(() => {});
    render(<FloatingWidget />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(scrollToTopSpy).toHaveBeenCalled();
    scrollToTopSpy.mockRestore();
  });
});
