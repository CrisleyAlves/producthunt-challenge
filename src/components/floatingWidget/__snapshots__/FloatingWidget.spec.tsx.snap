// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FloatingWidget > renders the floating widget button 1`] = `
<div>
  <button
    aria-label="Scroll to top"
    class="floating-widget"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-arrow-up icon"
      data-icon="arrow-up"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M214.6 41.4c-12.5-12.5-32.8-12.5-45.3 0l-160 160c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 141.2 160 448c0 17.7 14.3 32 32 32s32-14.3 32-32l0-306.7L329.4 246.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3l-160-160z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
